pkg/blockstorage/vmware/vmware.go

```diff
@@ -251,38 +253,54 @@ var reVslmSyncFaultFatal = regexp.MustCompile("Change tracking invalid or disk i
 // SnapshotCreate is part of blockstorage.Provider
 func (p *FcdProvider) SnapshotCreate(ctx context.Context, volume blockstorage.Volume, tags map[string]string) (*blockstorage.Snapshot, error) {
 	var res types.AnyType
+	description := generateDescription(tags)
 	err := wait.PollImmediate(time.Second, defaultRetryLimit, func() (bool, error) {
-		log.Debug().Print("CreateSnapshot", field.M{"VolumeID": volume.ID})
-		task, lerr := p.Gom.CreateSnapshot(ctx, vimID(volume.ID), noDescription)
-		if lerr != nil {
-			return false, errors.Wrap(lerr, "CreateSnapshot task creation failure")
+		createErr := func() error {
+			log.Debug().Print("CreateSnapshot", field.M{"VolumeID": volume.ID})
+			task, lerr := p.Gom.CreateSnapshot(ctx, vimID(volume.ID), description)
+			if lerr != nil {
+				return errors.Wrap(lerr, "CreateSnapshot task creation failure")
+			}
+			log.Debug().Print("Started CreateSnapshot task", field.M{"VolumeID": volume.ID})
+			res, lerr = task.Wait(ctx, vmWareTimeout)
+			if lerr != nil {
+				return lerr
+			}
+			return nil
+		}()
+
+		if createErr == nil {
+			return true, nil
 		}
-		log.Debug().Print("Started CreateSnapshot task", field.M{"VolumeID": volume.ID})
-		res, lerr = task.Wait(ctx, vmWareTimeout)
-		if lerr != nil {
-			if soap.IsVimFault(lerr) {
-				switch soap.ToVimFault(lerr).(type) {
-				case *types.InvalidState:
-					log.Error().WithError(lerr).Print("There is some operation, other than this CreateSnapshot invocation, on the VM attached still being protected by its VM state. Will retry", field.M{"VolumeID": volume.ID})
-					return false, nil
-				case *vslmtypes.VslmSyncFault: // potentially can leak snapshots
-					if vFault, ok := soap.ToVimFault(lerr).(*vslmtypes.VslmSyncFault); ok {
-						log.Error().Print(fmt.Sprintf("VslmSyncFault: %#v", vFault))
-					}
-					if !(govmomiError{lerr}).Matches(reVslmSyncFaultFatal) {
-						log.Error().Print(fmt.Sprintf("CreateSnapshot failed with VslmSyncFault. Will retry: %s", (govmomiError{lerr}).Format()), field.M{"VolumeID": volume.ID})
-						return false, nil
-					}
-					return false, errors.Wrap(lerr, "CreateSnapshot failed with VslmSyncFault. A snapshot may have been created by this failed operation")
-				case *types.NotFound:
-					log.Error().WithError(lerr).Print("CreateSnapshot failed with NotFound error. Will retry", field.M{"VolumeID": volume.ID})
+
+		// it's possible that snapshot was created despite of SOAP errors
+		res = p.getCreatedSnapshot(ctx, tags, volume.ID, time.Now().Add(-1*vmWareTimeout))
```

Comment:
Shouldn't you just save the time `epoch` when the function was called?

Reply:
Why pass in `tags`? Shouldn't the arguments just be:

- context
- description string
- volume ID
- epoch